service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
+}
 
-	// commandConn is the Connection the bridge receives commands (such as
-	// ComputeSystemCreate) over.
-	commandConn transport.Connection
+// Handler responds to a bridge request.
+type Handler interface {
+	ServeMsg(ResponseWriter, *Request)
+}
+
+// HandlerFunc is an adapter to use functions as handlers
+type HandlerFunc func(ResponseWriter, *Request)
+
+// ServeMsg calls f(w, r)
+func (f HandlerFunc) ServeMsg(w ResponseWriter, r *Request) {
+	f(w, r)
+}
+
+// Mux is a protocol multiplexer for request response pairs
+// following the bridge protocol.
+type Mux struct {
+	mu sync.Mutex
+	m  map[prot.MessageIdentifier]Handler
+}
 
-	// writeLock must be held while writing to the transport to ensure that
-	// message data is kept together when writing from multiple go routines
-	// simultaneously.
-	writeLock sync.Mutex
+// NewBridgeMux creates a default bridge multiplexer
+func NewBridgeMux() *Mux {
+	return new(Mux)
 }
 
-// NewBridge produces a new bridge struct using the given Transport and Core
-// interfaces.
-func NewBridge(tport transport.Transport, coreint core.Core) *bridge {
-	return &bridge{
-		tport:   tport,
-		coreint: coreint,
+// Handle registers the handler for the giveen message id
+func (mux *Mux) Handle(id prot.MessageIdentifier, handler Handler) {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+
+	if handler == nil {
+		panic("bridge: nil handler")
+	}
+
+	if mux.m == nil {
+		mux.m = make(map[prot.MessageIdentifier]Handler)
 	}
+
+	if _, ok := mux.m[id]; ok {
+		logrus.Infof("bridge: overwriting bridge handler for type: %d", id)
+	}
+
+	mux.m[id] = handler
 }
 
-// CommandLoop is the main body of the bridge client. It waits for messages
-// from the HCS, carries out the operations requested by the messages, responds
-// to the HCS, and repeats.
-func (b *bridge) CommandLoop() {
-	if err := b.loop(); err != nil {
-		logrus.Error(err)
+// HandleFunc registers the handler function for the given message id
+func (mux *Mux) HandleFunc(id prot.MessageIdentifier, handler func(ResponseWriter, *Request)) {
+	if handler == nil {
+		panic("bridge: nil handler func")
 	}
+
+	mux.Handle(id, HandlerFunc(handler))
 }
 
-func (b *bridge) loop() error {
-	var err error
-	b.commandConn, err = b.createAndConnectCommandConn()
-	if err != nil {
-		return err
+// Handler returns the handler to use for the given request type.
+func (mux *Mux) Handler(r *Request) Handler {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+
+	if r == nil {
+		panic("bridge: nil request to handler")
 	}
-	defer b.commandConn.Close()
 
-	for {
-		// Messages from the HCS come as a header, which contains information
-		// about the message (including its size), and a message body.
-		// The header and body contain the operation to be performed, as well
-		// as any information needed to perform the operation.
-		message, header, err := readMessage(b.commandConn)
-		if err != nil {
-			return errors.Wrap(err, "failed to read a message from the HCS")
-		}
+	if mux.m == nil {
```

Comment:
Ah, I see. It would be `nil` if no handlers have been registered yet. I think this is related to just initializing the map when `mux` is created. That removes the need for any special casing like this.